Change AVG aggregates for integer datatypes to return correct data types

shah-nirmit opened the pull request:
### Description
Problem:-
The AVG aggregates were having wrong return types for Integer datatypes

Fix Approach:-
New final functions were introduced to return correct datatype and  sys.avg(bigint),sys.avg(int),sys.avg(smallint),sys.avg(int) aggregates were created in babelfishpg_common extension  so that these aggregates are given more precedance than pg_catalog.avg when the dialect is `tsql` / called from tds endpoint

References 
[current-behaviour](https://www.postgresql.org/docs/current/functions-aggregate.html#:~:text=avg%20(%20smallint,bigint%20)%20%E2%86%92%20numeric)
[expected-behaviour](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/functions/avg-transact-sql.md#return-types)

#### Engine PR
https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/82

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

### Issues Resolved
BABEL-3507,BABEL-3672

### Tests scenarios covered
- Use case based
  - Tested the aggregates within functions procedures and views ,also with PARTITION and ORDER BY Clauses (Partial Aggregation)
- Boundary conditions
  - Tested for datatype overflow/underflow for int/bigint return types, errors match expected outputs
- Arbitrary inputs
  - Tested with arbitrary inputs , with PARTITION and OVER Clauses
- Minor version upgrade tests
  - Added dependency tests to see upgrade impact 
- Major version upgrade tests
  - Added dependency tests to all upgrade paths to check upgrade impact 



### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Add support for Avg Aggregates for Integer Datatypes

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Use the bigint_utility function for both SUM and AVG

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Add test cases for AVG aggregate

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3507-pg-15

Commit:
Add upgrade test and add avg aggregate to upgrade sql script

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Fixed upgrade test failures

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Fixed issue with upgrade script

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3507-pg-15

Commit:
Add upgrade sql script for 2.3-2.4

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

surendravishnoi commented on contrib/babelfishpg_common/sql/numerics.sql:
```diff
@@ -313,11 +313,23 @@ RETURNS BIGINT
 AS 'babelfishpg_common' , 'int4int2_sum'
 LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+CREATE OR REPLACE FUNCTION sys.int4int2_avg(pg_catalog._int8)
```
why pg_catalog._int8 and not BIGINT?

shah-nirmit replied:
The transition functions used for the aggregates (`int2_avg_accum`/`int4_avg_accum`) use `_int8` type as the state ,so this state is passed to the final functions; so the input to the final function is given as `pg_catalog._int8`

surendravishnoi replied:
please add Jungkook for review, we need to make sure this is not a problem

2jungkook replied:
It looks okay.

surendravishnoi commented on contrib/babelfishpg_common/src/numeric.c:
```diff
@@ -1037,7 +1039,7 @@ tsql_numeric_get_typmod(Numeric num)
 bigint_sum(PG_FUNCTION_ARGS)
 {
-	return bigint_poly_sum(fcinfo);
+	return bigint_utility(fcinfo, true);
```
I think it would be best to add an enum on engine side and use that. Boolean looks error prone.

shah-nirmit replied:
Resolved in the latest commit to engine PR

surendravishnoi commented on contrib/babelfishpg_common/src/numeric.c:
```diff
@@ -1065,3 +1067,38 @@ int4int2_sum(PG_FUNCTION_ARGS)
+	Int8TransTypeData *transdata;
+
+	if (ARR_HASNULL(transarray) ||
+			ARR_SIZE(transarray) != ARR_OVERHEAD_NONULLS(1) + sizeof(Int8TransTypeData))
```
how is this check validating 2 element array?

shah-nirmit replied:
This check is also being used in engine code for function `int8_avg` which is the final function that is used by `pg_catalog.avg` , so I have used the same validation logic for the state received .
Reference [link](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/BABEL_3_X_DEV__PG_15_X/src/backend/utils/adt/numeric.c#L6672-L6693)

surendravishnoi commented on contrib/babelfishpg_common/src/numeric.c:
```diff
@@ -1065,3 +1067,38 @@ int4int2_sum(PG_FUNCTION_ARGS)
+	if (transdata->count == 0)
+		PG_RETURN_NULL();
+	
+	if (unlikely(transdata->sum < PG_INT32_MIN) || unlikely(transdata->sum > PG_INT32_MAX))
```
put in a single unlikely

shah-nirmit replied:
Resolved in the latest commit

surendravishnoi replied:
it is still the same

surendravishnoi commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -223,6 +223,7 @@ Function sys.bbf_varbinary_cmp(sys.bbf_varbinary,sys.bbf_varbinary)
 Function sys.bbfbinary_sqlvariant(sys.bbf_binary)
 Function sys.bbfvarbinary(sys.bbf_varbinary,integer,boolean)
 Function sys.bbfvarbinary_sqlvariant(sys.bbf_varbinary)
+Function sys.bigint_avg(internal)
```
we are missing upgrade dependency checks for these

shah-nirmit replied:
There are C type function (final functions) that are being used when the aggregate are called . There are dependency objects(functions,procedures,views) in the test files. do we need to have dependency tests for them explicitly?

surendravishnoi replied:
Talk to Rishabh, see if we are taking in account indirect dependencies or not?

rishabhtanwar29 replied:
Currently we are not tracking indirect dependencies as even pg_depend does not track such dependencies. This looks like a function to function dependency where aggregate function depends upon other functions. But IMO, adding test for aggregate function should be enough since it's helper functions will anyway get executed/tested as part of it.

surendravishnoi commented on test/JDBC/input/BABEL-3507-vu-prepare.sql:
```diff
@@ -0,0 +1,112 @@
+CREATE TABLE babel_3507_vu_prepare_t1(
+avgbigint BIGINT, avgint INT , avgsmallint SMALLINT , avgtinyint TINYINT )
+GO
+
```
we should also add tests for negative/positive values mix, NULL values, divide by zero, result being zero

shah-nirmit replied:
Added extra tests in the latest commit.
`divide-by-zero` situation would not arise since if all values are NULL we don't increment the `state->count` . and return NULL if `state->count` is 0; [Reference](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1045/files#diff-ff0c89ab284f5bb75c83374ee361d8b192bca9db405599bb4bbcd316eddea1afR1090-R1091)

Deepesh125 commented on contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--2.3.0--2.4.0.sql:
```diff
@@ -0,0 +1,54 @@
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
```
why do we need to update this upgrade script if this change is only meant for 3_X_DEV branch?

shah-nirmit replied:
Fixed in latest commit, we need as to have all the 2_X sql upgrade script in 3_X dev as well

Deepesh125 commented on contrib/babelfishpg_common/src/numeric.c:
```diff
@@ -1065,3 +1067,38 @@ int4int2_sum(PG_FUNCTION_ARGS)
+
+	if (ARR_HASNULL(transarray) ||
+			ARR_SIZE(transarray) != ARR_OVERHEAD_NONULLS(1) + sizeof(Int8TransTypeData))
+			elog(ERROR, "expected 2-element int8 array");
```
use ereport instead of elog.

shah-nirmit replied:
Fixed in latest commit

Deepesh125 commented on test/JDBC/expected/BABEL-3507-vu-prepare.out:
```diff
@@ -0,0 +1,148 @@
+-- max values causing overflow in avg (BIGINT,INT) 
```
what would happen if one of the value is NULL?

shah-nirmit replied:
If one of the values are NULL , they are not used in avg calculation thus not causing overflow

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3507-pg-15

Commit:
Use enum for aggregate mode in utility function

Fixed use of ereport instead of elog

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

2jungkook commented on contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.0.0--3.1.0.sql:
```diff
@@ -0,0 +1,54 @@
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
+\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO '3.1.0'" to load this file. \quit
```
Please add `2.3.0--2.4.0` and `2.4.0--3.0.0` upgrade scripts.

shah-nirmit replied:
Fixed in the latest commit

2jungkook commented on test/JDBC/upgrade/13_6/schedule:
```diff
@@ -294,3 +294,4 @@ ISC-Table_Constraints
 sys_server_principals_dep_for_13_x
 sys_database_principals_dep_for_13_x
 case_insensitive_collation
+BABEL-3507-before-15-1
```
1. Please use a more descriptive test name other than BABEL-3507.
2. It should have `before-15-2-or-14-7` suffix because our current dev cycle is targeting PG15.2 and PG14.7 (cc: @rishabhtanwar29)
3. `BABEL-3507-before-15-1` and `BABEL-3507` has only one line difference. In that case, please consider creating three test files to avoid duplicates, e.g., `<test name>`, `<test name>-type-before-15-2-or-14-7` and `<test name>-type`.

shah-nirmit replied:
Fixed in the latest commit,
Divide existing test into common and Dependency tests

2jungkook commented on test/JDBC/upgrade/14_7/schedule:
```diff
@@ -365,5 +365,6 @@ sys_database_principals_dep
 atn2
 app_name
 case_insensitive_collation
+BABEL-3507-before-15-1
```
This should be `BABEL-3507`

shah-nirmit replied:
Fixed in the latest commit

Commit:
Divide test into common and dep

Added sql upgrade script for 2.3->2.4
Added sql upgrade script for 2.4->3.0
utitlity function name changed

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3507-pg-15

Commit:
Empty-Commit

Commit:
updated expected drop file

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

surendravishnoi commented on contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--2.4.0--3.0.0.sql:
```diff
@@ -0,0 +1,29 @@
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
```
why have we added collation related changes here?

shah-nirmit replied:
With accordance to https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1045#discussion_r1072608467
In case there is scenario where one upgrades first from 14.6->14.7 and then 14.7->15.2. 
We would need the collation changes that were also done for 2.3->3.0 . Both of the upgrade script should be identical

Deepesh125 replied:
Yes, this changes are required to make the MVU work.

surendravishnoi commented on test/JDBC/input/AVG-Aggregate-Dep-before-15-2-or-14-7-vu-prepare.sql:
```diff
@@ -0,0 +1,58 @@
+CREATE TABLE avg_dep_vu_prepare_t1(
```
why is this a separate file?

shah-nirmit replied:
We need the seperate file as the dependant objects (only dependant view) would see a change in return type returning (numeric for all avg aggregate before the fix). Since function and procedure are compiled on runtime they would have correct (new ) return types

Commit:
Fixed unlikely expression

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

rishabhtanwar29 commented:
Upgrade changes look good.

shah-nirmit reviewed (commented):
Note:- the 14_7->latest upgrade path will fail for AVG-Aggregate-Dep ,2X PR needs to be merged First

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3507-pg-15
